aesara/tensor/elemwise.py

```diff
@@ -1351,26 +1350,20 @@ def make_node(self, input):
                     f"Not enough dimensions on {input} to reduce on axis {a}"
                 )
             if a < 0:
-                copy_op = True
                 axis[i] = a + inp_dims
 
-        # We can't call self.__class__() as there is a class that
-        # inherits from CAReduce that doesn't have the same signature
-        if copy_op:
-            op = copy(self)
-            op.set_ufunc(op.scalar_op)
-            assert len(axis) == len(self.axis)
-            op.axis = tuple(axis)
-        else:
-            op = self
-
         broadcastable = [x for i, x in enumerate(inp_bdcast) if i not in axis]
 
         output = TensorType(
             dtype=self._output_dtype(inp_dtype), broadcastable=broadcastable
         )()
 
-        return Apply(op, [input], [output])
+        if isinstance(axis, int):
+            axis = as_tensor_variable([axis], ndim=1)
+        else:
+            axis = as_tensor_variable(axis, ndim=1)
+
+        return Apply(self, [input, axis], [output])
```

Comment:
We need to remove `self.axis` and add `axis` to `make_node`'s signature.  

In general, `make_node`'s signature should match the `Apply` node's inputs.  This correspondence between `Apply.inputs` and `Op.make_node`'s signature makes it possible for us to easily recreate a node using `Op.make_node(*Apply.inputs)`.  If anything, it would be great if we could confirm/guarantee this correspondence during class creation (e.g. in a metaclass `__new__`).

N.B.: This was the exact issue I fixed in #344.

Reply:
Looks like `TensorConstant` isn't allowed in the `make_node` argument list. It is only permitting symbolic inputs like `TensorVariable`. Is there any way we could get the `axis` data (and not just it's `shape` and `dtype`) into `make_node` so that output's broadcasting shape is calculated properly ?

Reply:
> Looks like `TensorConstant` isn't allowed in the `make_node` argument list. It is only permitting symbolic inputs like `TensorVariable`.

That's weird; what was the exact error?

Reply:
In the `with_mode` method of `test_elemwise.TestCAReduce`, I passed the `tosum` parameter (which is basically the `axis`) as a `TensorConstant` while building of the `function`. 

>  f = aesara.function([x, as_tensor_variable(tosum)], e, mode=mode)

It throws with the following stack trace : 

```tests/tensor/test_elemwise.py:357: in with_mode
    f = aesara.function([x, as_tensor_variable(tosum)], e, mode=mode)
aesara/compile/function/__init__.py:337: in function
    fn = pfunc(
aesara/compile/function/pfunc.py:426: in pfunc
    inputs = [
aesara/compile/function/pfunc.py:427: in <listcomp>
    _pfunc_param_to_in(p, allow_downcast=allow_input_downcast) for p in params
_ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ 

param = TensorConstant{[0 1]}, strict = False, allow_downcast = None

    def _pfunc_param_to_in(param, strict=False, allow_downcast=None):
        if isinstance(param, Constant):
>           raise TypeError("Constants not allowed in param list", param)
E           TypeError: ('Constants not allowed in param list', TensorConstant{[0 1]})```

Reply:
I just tried these tests locally and I don't see that issue; however, it looks like the axes aren't being summed, so the tests are failing.

Reply:
@brandonwillard Can you please elaborate how the test passed locally. (What exactly did you try ?). Was the `axis` passed as `TensorConstant` ?

> however, it looks like the axes aren't being summed, so the tests are failing.

In the current implementation, yes, since the axes are being passed as `TensorVariables`. I don't have their data in the `make_node` or `infer_shape` methods. My initial doubt is exactly that - How to get that data inside these methods, since `Tensorvariable` seems to give information only about the shape and datatype of the object.